vim-engine/src/main/kotlin/com/maddyhome/idea/vim/api/VimMarkServiceBase.kt

```diff
@@ -0,0 +1,605 @@
+    setMark(caret, SELECTION_START_MARK, range.startOffset)
+    setMark(caret, SELECTION_END_MARK, range.endOffset)
+  }
+
+  override fun getVisualSelectionMarks(caret: ImmutableVimCaret): TextRange? {
+    return getMarksRange(caret, SELECTION_START_MARK, SELECTION_END_MARK)
+  }
+
+  override fun setChangeMarks(caret: ImmutableVimCaret, range: TextRange) {
+    setMark(caret, CHANGE_START_MARK, range.startOffset)
+    setMark(caret, CHANGE_END_MARK, range.endOffset - 1)
+  }
+
+  override fun getChangeMarks(caret: ImmutableVimCaret): TextRange? {
+    return getMarksRange(caret, CHANGE_START_MARK, CHANGE_END_MARK)
+  }
+
+  private fun getMarksRange(caret: ImmutableVimCaret, startMark: Char, endMark: Char): TextRange? {
+    val editor = caret.editor
+    val startOffset = getLocalMark(caret, startMark)
+      ?.offset(editor) ?: return null
+    val endOffset = getLocalMark(caret, endMark)
+      ?.offset(editor) ?: return null
+    return TextRange(startOffset, endOffset + 1)
+  }
+
+  override fun removeMark(editor: VimEditor, char: Char) {
+    val markChar = if (char == '`') '\'' else char
+    if (markChar.isLocalMark()) {
+      editor.carets()
+        .forEach { removeLocalMark(it, char) }
+    }
+    if (markChar.isGlobalMark()) {
+      removeGlobalMark(char)
+    }
+  }
+
+  override fun removeLocalMark(caret: ImmutableVimCaret, char: Char) {
+    val markChar = if (char == '`') '\'' else char
+    if (!markChar.isLocalMark()) return
+
+    if (caret.isPrimary) {
+      if (markChar == SELECTION_START_MARK) {
+        val selectionInfo = caret.lastSelectionInfo
+        val startPosition = selectionInfo.start
+        if (startPosition != null) caret.lastSelectionInfo = SelectionInfo(null, selectionInfo.end, selectionInfo.type)
+      } else if (markChar == SELECTION_END_MARK) {
+        val selectionInfo = caret.lastSelectionInfo
+        val endPosition = selectionInfo.end
+        if (endPosition != null) caret.lastSelectionInfo = SelectionInfo(selectionInfo.start, null, selectionInfo.type)
+      } else {
+        val path = caret.editor.getPath() ?: return
+        getLocalMarks(path).remove(markChar)
+      }
+    } else {
+      caret.markStorage.removeMark(markChar, caret)
+    }
+  }
+
+  override fun removeGlobalMark(char: Char) {
+    val markChar = if (char == '`') '\'' else char
+    if (markChar.isGlobalMark()) {
+      globalMarks.remove(markChar)
+    }
+  }
+
+  override fun updateMarksFromInsert(editor: VimEditor, insertStartOffset: Int, insertLength: Int) {
+    val caretToMarks = getAllMarksForFile(editor)
+    if (caretToMarks.isEmpty()) return
+
+    val insertEndOffset = insertStartOffset + insertLength
+    val insStart = editor.offsetToBufferPosition(insertStartOffset)
+    val insEnd = editor.offsetToBufferPosition(insertEndOffset)
+    logger.debug { "mark insert. insStart = $insertStartOffset, insEnd = $insertEndOffset" }
+    val lines = insEnd.line - insStart.line
+    if (lines == 0) return
```

Comment:
Why don't we update marks if the amount of lines was not changed? What about the colum-dependent marks?